google/cloud/security/scanner/audit/bigquery_rules_engine.py

```diff
@@ -0,0 +1,284 @@
+
+    To require > 0 length glob, change the "*" to ".+". This is to handle
+    strings like "*@company.com". (THe actual regex would probably be
+    ".*@company.com", except that we don't want to match zero-length
+    usernames before the "@".)
+
+    Args:
+        pattern_string (str): The pattern string of which to make a regex.
+
+    Returns:
+        str: The pattern string, escaped except for the "*", which is
+        transformed into ".+" (match on one or more characters).
+    """
+
+    return '^{}$'.format(re.escape(pattern_string).replace('\\*', '.+'))
+
+
+class BigqueryRulesEngine(bre.BaseRulesEngine):
+    """Rules engine for Big Query data sets"""
+
+    def __init__(self, rules_file_path, snapshot_timestamp=None):
+        """Initialize.
+
+        Args:
+            rules_file_path (str): file location of rules
+            snapshot_timestamp (str): snapshot timestamp. Defaults to None.
+                If set, this will be the snapshot timestamp
+                used in the engine.
+        """
+        super(BigqueryRulesEngine,
+              self).__init__(rules_file_path=rules_file_path)
+        self.rule_book = None
+
+    def build_rule_book(self):
+        """Build BigqueryRuleBook from the rules definition file."""
+        self.rule_book = BigqueryRuleBook(self._load_rule_definitions())
+
+    # pylint: disable=arguments-differ
+    def find_policy_violations(self, bq_datasets,
+                               force_rebuild=False):
+        """Determine whether Big Query datasets violate rules.
+
+        Args:
+            bq_datasets (list): Object containing ACL
+                data
+            force_rebuild (bool): If True, rebuilds the rule book. This will
+                reload the rules definition file and add the rules to the book.
+
+        Returns:
+             generator: A generator of rule violations.
+        """
+        violations = itertools.chain()
+        if self.rule_book is None or force_rebuild:
+            self.build_rule_book()
+        resource_rules = self.rule_book.get_resource_rules()
+
+        for rule in resource_rules:
+            violations = itertools.chain(violations,
+                                         rule.\
+                                         find_policy_violations(bq_datasets))
+        return violations
+
+    def add_rules(self, rules):
+        """Add rules to the rule book.
+
+        Args:
+            rules (dict): rule definitions dictionary
+        """
+        if self.rule_book is not None:
+            self.rule_book.add_rules(rules)
+
+
+class BigqueryRuleBook(bre.BaseRuleBook):
+    """The RuleBook for Big Query dataset resources."""
+
+    def __init__(self, rule_defs=None):
+        """Initialization.
+
+        Args:
+            rule_defs (dict): rule definitons dictionary
+        """
+        super(BigqueryRuleBook, self).__init__()
+        self.resource_rules_map = {}
+        if not rule_defs:
+            self.rule_defs = {}
```

Comment:
The reason the original code was written with the default to None is because of this weird Python gotcha: http://python-guide-pt-br.readthedocs.io/en/latest/writing/gotchas/